mycore-ocfl/src/main/java/org/mycore/ocfl/metadata/MCROCFLXMLMetadataManager.java

```diff
@@ -317,7 +328,34 @@ public IntStream getStoredIDs(String project, String type) throws MCRPersistence
 
     @Override
     public int getHighestStoredID(String project, String type) {
-        return getStoredIDs(project, type).max().orElse(0);
+        Path path
+            = Path.of(MCRConfiguration2.getStringOrThrow("MCR.OCFL.Repository." + repositoryKey + ".RepositoryRoot"));
+        int highestStoredID = 0;
+        int maxDepth = Integer.MAX_VALUE;
+        OcflExtensionConfig config = MCRConfiguration2.getSingleInstanceOf("MCR.OCFL.Repository." + repositoryKey)
+            .map(MCROCFLRepositoryProvider.class::cast).orElseThrow().getExtensionConfig();
+        // optimization for known layouts
+        if (Objects.equals(config.getExtensionName(), MCRStorageLayoutExtension.EXTENSION_NAME)) {
+            maxDepth = ((MCRStorageLayoutConfig) config).getSlotLayout().split("-").length;
+            path = Path.of(path.toString(), MCROCFLObjectIDPrefixHelper.MCROBJECT.replace(":", ""), project, type);
+        } else if (Objects.equals(config.getExtensionName(),
+            HashedNTupleIdEncapsulationLayoutExtension.EXTENSION_NAME)) {
+            maxDepth = ((HashedNTupleIdEncapsulationLayoutConfig) config).getNumberOfTuples() + 1;
+        }
+        try (Stream<Path> stream
+            = Files.find(path, maxDepth, (filePath, fileAttr) -> filePath.getFileName().toString()
+                .matches("^.*" + project + "_{1}" + type + "_{1}\\d+$"))) {
+            highestStoredID = stream.map(entry -> entry.getFileName().toString())
+                .map(fileName -> Integer.parseInt(fileName.substring(fileName.lastIndexOf('_') + 1)))
+                .max(Comparator.comparing(Integer::valueOf)).orElse(0);
```

Comment:
This could be simplified to   `.max(Integer::compareTo)`

Reply:
If we can assume that the number_length of the identifier are equal / do not change in the whole OCFL repository, we could  compare on Path instead of Integer and do the `.max()` earlier in the stream:
```
try (Stream<Path> stream
            = Files.find(path, maxDepth, (filePath, fileAttr) -> filePath.getFileName().toString().startsWith(TEST_PREFIX))) {
            return stream
                .map(p -> p.getFileName())
                .max(Path::compareTo)
                .map(entry -> entry.getFileName().toString())
                .map(smax -> Integer.parseInt(smax.substring(smax.lastIndexOf('_') + 1)))
                .orElse(0);
         }
```
This brings only a very little performance, but may make the algorithm more understandable.